### ISSUE TITLE
Implement parameters expansion

### ISSUE THREAD

**Issue by piranna** [vorpaljs/bash-parser#19]
It should return AST nodes and not try to expand it, at least on a first iteration.

**Comment by parro-it** [comment]
I already started to work on simple parameter expansion. Parameters are identified in [the tokenizer](https://github.com/vorpaljs/bash-parser/blob/master/src/token-delimiter.js#L115) and there are some test [here](https://github.com/vorpaljs/bash-parser/blob/master/test/test-lexer.js#L23).

For this source: `echo word${other}test` we actually emit these two tokens:
```js
[
                {
                         token: 'WORD', value: 'echo'
                },
		{
			token: 'WORD',
			value: 'word${other}test',
			expansion: {text: 'other', start: 4, end: 12}
		} 
]
```

The `expansion` node should have sufficient informations for the user to do the expansion by itself. Parameters does not have the intricacies of commands and alias, so this should be relatevely easy to complete.

Still to do:

* Pass the `expansion` node from the token to the AST
* Parse all the parameters operators:

> ${parameter:-[word]}
> Use Default Values. If parameter is unset or null, the expansion of word (or an empty string if word is omitted) shall be substituted; otherwise, the value of parameter shall be substituted.
> ${parameter:=[word]}
> Assign Default Values. If parameter is unset or null, the expansion of word (or an empty string if word is omitted) shall be assigned to parameter. In all cases, the final value of parameter shall be substituted. Only variables, not positional parameters or special parameters, can be assigned in this way.
> ${parameter:?[word]}
> Indicate Error if Null or Unset. If parameter is unset or null, the expansion of word (or a message indicating it is unset if word is omitted) shall be written to standard error and the shell exits with a non-zero exit status. Otherwise, the value of parameter shall be substituted. An interactive shell need not exit.
> ${parameter:+[word]}
> Use Alternative Value. If parameter is unset or null, null shall be substituted; otherwise, the expansion of word (or an empty string if word is omitted) shall be substituted.
>

**Comment by piranna** [comment]
Add a test when the expansion is alone without surrounding text, just to be sure. Also we should think about emiting `command` tokens instead of generic `WORD`... Maybe this should be task of another AST parser, and left this to be canonical to the spec, giving its output AST as input for the other, higher level one? It would makes sense to add the parsing of `ASSIGMENT_WORD` tokens to this expanded one... How do you see it?

**Comment by parro-it** [comment]
> Add a test when the expansion is alone without surrounding text, just to be sure.
:+1: 

All type of expansion (Parameters, Command, Arithmetic) should happens inside any WORD token.
`ASSIGMENT_WORD` tokens identification always happen AFTER expansions, so `ASSIGMENT_WORD` tokens are already covered.

**Comment by parro-it** [comment]
The tokenizer is actually implemented as a set of composed generators, where each one take the output iterator of the preceding as its input.

The most complex one is the first, that take as input the raw text source code and emit 
tokens based on the rules specified in chapter "2.10.2 Shell Grammar Rules" of the spec.

The other ones each implement a particular aspact of the tokinzation phase.
You can see [here](https://github.com/vorpaljs/bash-parser/blob/master/src/posix-shell-lexer.js) how they are composed to form the jison lexer object.

**Comment by piranna** [comment]
I was asking it just to make clear what the spec says and what were our additions.

**Comment by parro-it** [comment]
:+1: I'm trying to explain how the actual implementation work just to share the knowledge...

**Comment by parro-it** [comment]
> Maybe this should be task of another AST parser, and left this to be canonical to the spec, giving its output AST as input for the other, higher level one?

Please expand this concept, I'm not sure to understand it, seems somethink like the iterators composition I talk above

**Comment by piranna** [comment]
Sort of, one parser that emits tokens as defined by the spec, and another
expanded one based on it that emits the expanded tokens.

**Comment by parro-it** [comment]
Nice idea! The second one could be a kind of visitor on the AST resulting from the first...

**Comment by parro-it** [comment]
I just pushed a version that pass parameters substitution up to the final AST. A passing test is [here](https://github.com/vorpaljs/bash-parser/blob/arg-subst/test/test-ast.js#L6)

**Comment by parro-it** [comment]
That test is passing, but I created a separate branch because now it break many test due to the change in argument AST.

**Comment by parro-it** [comment]
What do you think of the provided expansion info? Do you think they make it easy to handle the substitution on user side?

**Comment by piranna** [comment]
> Nice idea! The second one could be a kind of visitor on the AST resulting from the first...

That's the idea. Other option would be that the second one piggy-back on top of the spec one or inherit over it and emit directly the AST nodes.


> What do you think of the provided expansion info? Do you think they make it easy to handle the substitution on user side?

Seems some of the words are missing, isn't it?

**Comment by parro-it** [comment]
> Seems some of the words are missing, isn't it?

Sorry I don't understand, , what do you mean exactly?

**Comment by piranna** [comment]
The AST on the test only have the parameter expansion, but don't have the surrounding words in any place.

**Comment by piranna** [comment]
Or are you planning it should be replaced using `splice()`? In that case, the end should be one less, to define just the character positions where the substitution should be done (12 is the start of the word after the parameter expansion...).

**Comment by parro-it** [comment]
Ok, now I understood what you mean.
The plan is to simplify the replacement if using `String.slice`:

Given this AST:

```
var ast = {
    text: 'echoword=${other}test',
    expansion: [{
        text: 'other',
        start: 9,
        end: 17
};
```

You can do the replacement with this code:

```js
var exp = ast.expansion[0];
ast.text.slice(ast.text.slice(0, exp.start) + env[exp.text] +  ast.text.slice(exp.end))
```

`end` represent the first character __after__ the substitution. 
The name is really unhappy, how can we call it? "next"?

**Comment by parro-it** [comment]
I also thought of use the same representation that we have for es6 template strings:

```js
var ast = {
    text: ['echoword=', 'other', 'test'],
};
```

Even position in a WORD array represent strings, odd position represent substitution...

The code to do substitution is even simpler:

```js
var exp = ast.expansion[0];
ast.text[0] + env[ast.text[1]] +  ast.text[2]
```

But I think in this case the intent of the code is less clear.
Moreover, I plan to introduce a 'type' property on `expansion`, to distinguish parameters expansion from arithmetic and command ones....

**Comment by piranna** [comment]
I thought there was a `String.splice()` similar to `Array.splice()`, sorry.

> end represent the first character after the substitution.
> The name is really unhappy, how can we call it? "next"?

Yes, `next` is a better name, since it's the index of the next token or the end/length of the string.


> I also thought of use the same representation that we have for es6 template strings

I didn't though about template strings, but it makes sense, the format is the same both for shell and javascript (done in purposse?) and substitution would be trivial :-) The bad thing that it uses variable names in the current context and seems it's not easy to interpolate them... :-/

**Comment by parro-it** [comment]
> Yes, next is a better name, since it's the index of the next token or the end/length of the string.
:+1: I will change that

> I didn't though about template strings, but it makes sense, the format is the same both for shell and javascript (done in purposse?) and substitution would be trivial :-) The bad thing that it uses variable names in the current context and seems it's not easy to interpolate them... :-/

I would keep it as it is for now (I'm close to merging the branch in master). But keep this change in our mind, it could be a big simplification on user side!

**Comment by piranna** [comment]
> I would keep it as it is for now (I'm close to merging the branch in master). But keep this change in our mind, it could be a big simplification on user side!

:+1:

**Comment by parro-it** [comment]
I merged the branch into master. 
I do a refactor of all WORD and ASSIGMENT_WORD  AST that now are always represented as objects, while they are before represented as simple string.

This change is due to the fact that now these two tokens can contain an additional expansion property containing expansions present in the token.

**Comment by parro-it** [comment]
Tests for parameter expansion are on [test-parameter-substitution.js](https://github.com/vorpaljs/bash-parser/blob/master/test/test-parameter-substitution.js).

Also, when we have expansion operators such as in `${varname:-default_value}`
the spec says that default_value should have further expansion applyed, I call a recursive parse [here](https://github.com/vorpaljs/bash-parser/blob/master/src/token-delimiter.js#L60).
This is weird and I think it could break in many case, so it should be refactored to somethink more robust.

**Comment by piranna** [comment]
What an awesome work, good job! :-D

**Comment by parro-it** [comment]
Thank you! :smile_cat:

**Comment by piranna** [comment]
> Also, when we have expansion operators such as in ${varname:-default_value}
> the spec says that default_value should have further expansion applyed, I call a recursive parse here.
> This is weird and I think it could break in many case, so it should be refactored to somethink more robust.

What about just calling `parse()` and return a new AST subtree, similar to what it's done with parameters expansion?

**Comment by parro-it** [comment]
> What about just calling parse() and return a new AST subtree, similar to what it's done with parameters expansion? 

That's what I'm doing now, the problem is that the text should be always considered a single WORD token, and reparsing it this is not assured. I have to try to wrap it in double quotes or somethink...

**Comment by parro-it** [comment]
Other than the improvement above, there is also still to implements:

* Positional Parameters
* Special Parameters

**Comment by piranna** [comment]
> I have to try to wrap it in double quotes or somethink...

Maybe directly calling some of the internal parsers used after the tokenizer?

**Comment by parro-it** [comment]
I have extracted the parameters expansion logic to its own module, removing the fragile code we are talking above. The resulting module could also be used as a blueprint for the other kind of substitutions...

**Comment by piranna** [comment]
> I have extracted the parameters expansion logic to its own module

Good idea :+1:

**Comment by parro-it** [comment]
Implemented 

* Positional Parameters
* Special Parameters

Closing this issue.

**Comment by parro-it** [comment]
Published npm version 0.2.0 containing parameter expansion

**Comment by piranna** [comment]
Great, thank you! :-)
